backend/internal/db/migrations/003_add_indexes.sql

```diff
@@ -0,0 +1,65 @@
+-- Migration: Add performance indexes
+-- Description: Create indexes to optimize common queries
+
+-- Nodes table indexes
+-- Index for node type filtering
+CREATE INDEX IF NOT EXISTS idx_nodes_node_type ON nodes(node_type);
+
+-- Index for tags (using GIN for array operations)
+CREATE INDEX IF NOT EXISTS idx_nodes_tags ON nodes USING GIN(tags);
+
+-- Index for file path lookups
+CREATE INDEX IF NOT EXISTS idx_nodes_file_path ON nodes(file_path);
+
+-- Full-text search index on title and content
+CREATE INDEX IF NOT EXISTS idx_nodes_search ON nodes 
+USING GIN(to_tsvector('english', COALESCE(title, '') || ' ' || COALESCE(content, '')));
+
+-- Index for centrality-based sorting (for finding important nodes)
+CREATE INDEX IF NOT EXISTS idx_nodes_centrality ON nodes(centrality DESC);
+
+-- Composite index for common queries
+CREATE INDEX IF NOT EXISTS idx_nodes_type_centrality ON nodes(node_type, centrality DESC);
+
+-- Edges table indexes
+-- Index for source node lookups
+CREATE INDEX IF NOT EXISTS idx_edges_source_id ON edges(source_id);
+
+-- Index for target node lookups
+CREATE INDEX IF NOT EXISTS idx_edges_target_id ON edges(target_id);
```

Comment:
Consider adding a composite index for the common query pattern in `GetByNode` method which uses `WHERE source_id = $1 OR target_id = $1`. PostgreSQL can't efficiently use separate indexes for OR conditions. You might want to consider:

1. Using a UNION query instead (which can use both indexes):
```sql
SELECT * FROM edges WHERE source_id = $1
UNION
SELECT * FROM edges WHERE target_id = $1
```

2. Or adding a specialized index if this query is performance-critical at 50k scale.